lib/pronto/formatter/pull_request_formatter.rb

```diff
@@ -0,0 +1,87 @@
+module Pronto
+  module Formatter
+    class PullRequestFormatter
+      def format(messages, repo, patches)
+        client    = client_module.new(repo)
+        head      = repo.head_commit_sha
+        existing  = existing_comments(client, head)
+        comments  = new_comments(messages, patches, head)
+        additions = remove_duplicate_comments(existing, comments)
+        submit_comments(client, additions)
+
+        "#{additions.count} Pronto messages posted to #{pretty_name}"
+      end
+
+      def client_module
+        raise NotImplementedError
+      end
+
+      def pretty_name
+        raise NotImplementedError
+      end
+
+      private
+
+      def config
+        @config ||= Config.new
+      end
+
+      def consolidate_comments(comments)
+        result = comments.first
+        if comments.length > 1
+          joined_body = comments
+                        .map { |comment| "- #{comment.body}" }
+                        .join("\n")
+
+          result = client_module::Comment.new(result.sha, joined_body, result.path, result.position)
+        end
+        result
+      end
+
+      def existing_comments(client, sha)
+        client
+          .pull_comments(sha)
+          .group_by { |comment| [comment.path, comment.position] }
+      end
+
+      def new_comment(message, patches, sha)
+        body = message.msg
+        path = message.path
+        lineno = message.line.line.new_lineno
+
+        client_module::Comment.new(sha, body, path, lineno)
+      end
+
+      def new_comments(messages, patches, sha)
+        messages
+          .uniq
+          .map { |message| new_comment(message, patches, sha) }
+          .group_by { |comment| [comment.path, comment.position] }
+      end
+
+      def remove_duplicate_comments(old_comments, new_comments)
+        new_comments.each_with_object([]) do |(key, comments), memo|
+          existing = old_comments[key]
+          if existing
+            body     = existing.map(&:body).join(' ')
+            comments = comments.select { |comment| !body.include?(comment.body) }
+          end
+
+          if config.consolidate_comments?
+            comment = consolidate_comments(comments)
+            memo.push(comment)
+          else
+            memo.concat(comments)
+          end
+        end
+      end
+
+      def submit_comments(client, comments)
+        puts comments
+        comments.each { |comment| client.create_pull_comment(comment) }
+      rescue Octokit::UnprocessableEntity => e
```

Comment:
@doomspork `BitbucketClient` uses `httparty`. We could listen for `HTTParty::Error`.